src/map/clan.c

```diff
@@ -0,0 +1,873 @@
+{
+	struct DBIterator *iter = db_iterator(clan->db);
+	struct clan *c;
+
+	for (c = dbi_first(iter); dbi_exists(iter); c = dbi_next(iter)) {
+		script->set_constant2(c->constant, c->clan_id, false, false);
+	}
+
+	dbi_destroy(iter);
+
+	clan->fix_relationships();
+}
+
+/**
+ * Fix clan relationships (allies and antagonists)
+ */
+void clan_fix_relationships(void)
+{
+	struct DBIterator *iter = db_iterator(clan->db);
+	struct clan *c;
+	int i, clan_id;
+
+	for (c = dbi_first(iter); dbi_exists(iter); c = dbi_next(iter)) {
+		for (i = 0; i < ARRAYLENGTH(c->allies); i++) {
+			if (script->get_constant(c->allies[i].constant, &clan_id)) {
+				c->allies[i].clan_id = clan_id;
+			} else {
+				memset(&c->allies[i].constant, 0, NAME_LENGTH);
+				c->allies[i].clan_id = 0;
+			}
+		}
+		for (i = 0; i < ARRAYLENGTH(c->antagonists); i++) {
+			if (script->get_constant(c->antagonists[i].constant, &clan_id)) {
+				c->antagonists[i].clan_id = clan_id;
+			} else {
+				memset(&c->antagonists[i].constant, 0, NAME_LENGTH);
+				c->antagonists[i].clan_id = 0;
+			}
+		}
+	}
+	dbi_destroy(iter);
+}
+
+/**
+ * Returns the clan_id of bl
+ */
+int clan_get_id(struct block_list *bl)
+{
+	nullpo_ret(bl);
+	switch( bl->type ) {
+		case BL_PC: {
+			const struct map_session_data *sd = BL_UCCAST(BL_PC, bl);
+			return sd->status.clan_id;
+		}
+		case BL_PET: {
+			const struct pet_data *pd = BL_UCCAST(BL_PET, bl);
+			if (pd->msd != NULL)
+				return pd->msd->status.clan_id;
+		}
+			break;
+		case BL_MOB: {
+			const struct mob_data *md = BL_UCCAST(BL_MOB, bl);
+			const struct map_session_data *msd;
+			if (md->special_state.ai != AI_NONE && (msd = map->id2sd(md->master_id)) != NULL) {
+				return msd->status.clan_id;
+			}
+			return md->clan_id;
+		}
+		case BL_HOM: {
+			const struct homun_data *hd = BL_UCCAST(BL_HOM, bl);
+			if (hd->master != NULL) {
+				return hd->master->status.clan_id;
+			}
+		}
+			break;
+		case BL_MER: {
+			const struct mercenary_data *md = BL_UCCAST(BL_MER, bl);
+			if (md->master != NULL) {
+				return md->master->status.clan_id;
+			}
+		}
+			break;
+		case BL_SKILL: {
+			const struct skill_unit *su = BL_UCCAST(BL_SKILL, bl);
+			if (su->group != NULL)
+				return su->group->clan_id;
+		}
+			break;
+	}
+
+	return 0;
+}
+
+/**
+ * Checks every clan player and remove those who are inactive
+ */
+int clan_inactivity_kick(int tid, int64 tick, int id, intptr_t data)
+{
+	struct clan *c = NULL;
+	int i;
+
+	if ((c = clan->search(id)) != NULL) {
+		if (!c->kick_time || c->tid != tid || tid == INVALID_TIMER || c->tid == INVALID_TIMER) {
+		  ShowError("Timer Mismatch (Time: %d seconds) %d != %d", c->kick_time, c->tid, tid);
+		  return 0;
+		}
+		for (i = 0; i < ARRAYLENGTH(c->member); i++) {
+			if (c->member[i].char_id <= 0 || c->member[i].online <= 0)
+				continue;
+
+			if (c->member[i].online) {
+				struct map_session_data *sd = c->member[i].sd;
+				if (DIFF_TICK32(sockt->last_tick, sd->idletime) > c->kick_time) {
```

Comment:
This is to support custom clans with lower kick time, so afk players don't hold vacancies.